src/main/java/com/google/firebase/remoteconfig/ConditionEvaluator.java

```diff
@@ -0,0 +1,346 @@
+    try {
+      return Pattern.compile(targetSignal).matcher(customSignal).matches();
+    } catch (PatternSyntaxException e) {
+      return false;
+    }
+  }
+
+  private boolean compareNumbers(
+      ImmutableList<String> targetValues, String customSignal, IntPredicate compareFunction) {
+    if (targetValues.size() != 1) {
+      logger.warn(
+          String.format(
+              "Target values must contain 1 element for numeric operations. Target Value: %s",
+              targetValues));
+      return false;
+    }
+
+    try {
+      double customSignalDouble = Double.parseDouble(customSignal);
+      double targetValue = Double.parseDouble(targetValues.get(0));
+      int comparisonResult = Double.compare(customSignalDouble, targetValue);
+      return compareFunction.test(comparisonResult);
+    } catch (NumberFormatException e) {
+      logger.warn(
+          "Error parsing numeric values: customSignal=%s, targetValue=%s",
+          customSignal, targetValues.get(0), e);
+      return false;
+    }
+  }
+
+  private boolean compareSemanticVersions(
+      ImmutableList<String> targetValues, String customSignal, IntPredicate compareFunction) {
+    if (targetValues.size() != 1) {
+      logger.warn(String.format("Target values must contain 1 element for semantic operation."));
+      return false;
+    }
+
+    String targetValueString = targetValues.get(0);
+    if (!validateSemanticVersion(targetValueString) || !validateSemanticVersion(customSignal)) {
+      return false;
+    }
+
+    List<Integer> targetVersion = parseSemanticVersion(targetValueString);
+    List<Integer> customSignalVersion = parseSemanticVersion(customSignal);
+
+    int maxLength = 5;
+    if (targetVersion.size() > maxLength || customSignalVersion.size() > maxLength) {
+      logger.warn(
+          "Semantic version max length(%s) exceeded. Target: %s, Custom Signal: %s",
+          maxLength, targetValueString, customSignal);
+      return false;
+    }
+
+    int comparison = compareSemanticVersions(customSignalVersion, targetVersion);
+    return compareFunction.test(comparison);
+  }
+
+  private int compareSemanticVersions(List<Integer> version1, List<Integer> version2) {
+    int maxLength = Math.max(version1.size(), version2.size());
+    int version1Size = version1.size();
+    int version2Size = version2.size();
+
+    for (int i = 0; i < maxLength; i++) {
+      // Default to 0 if segment is missing
+      int v1 = i < version1Size ? version1.get(i) : 0;
+      int v2 = i < version2Size ? version2.get(i) : 0;
+
+      int comparison = Integer.compare(v1, v2);
+      if (comparison != 0) {
+        return comparison;
+      }
+    }
+    // Versions are equal
+    return 0;
+  }
+
+  private List<Integer> parseSemanticVersion(String versionString) {
+    return Arrays.stream(versionString.split("\\."))
+        .map(Integer::parseInt)
+        .collect(Collectors.toList());
+  }
+
+  private boolean validateSemanticVersion(String version) {
```

Comment:
Depends on how often this helper function will be called in runtime. We could optimize it by caching the result from the RegEx compilation.

```java
private static final Pattern SEMVER_PATTERN =  Pattern.compile("^[0-9]+(?:\\.[0-9]+){0,4}$");

private boolean validateSemanticVersion(String version) {
  return SEMVER_PATTERN.matcher(version).matches();
}
```